registry/registry.go

```diff
@@ -17,61 +17,126 @@ limitations under the License.
 package registry
 
 import (
-	"strings"
-
 	"github.com/kubernetes/deployment-manager/common"
+
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
 )
 
-type RegistryService interface {
-	// List all the registries
-	List() ([]*common.Registry, error)
-	// Create a new registry
-	Create(repository *common.Registry) error
-	// Get a registry
-	Get(name string) (*common.Registry, error)
-	// Delete a registry
-	Delete(name string) error
-	// Find a registry that backs the given URL
-	GetByURL(URL string) (*common.Registry, error)
+// Registry abstracts a registry that holds charts, which can be
+// used in a Deployment Manager configuration. There can be multiple
+// registry implementations.
+type Registry interface {
+	// GetRegistryName returns the name of this registry
+	GetRegistryName() string
+	// GetRegistryType returns the type of this registry.
+	GetRegistryType() common.RegistryType
+	// GetRegistryShortURL returns the short URL for this registry.
+	GetRegistryShortURL() string
+	// GetRegistryFormat returns the format of this registry.
+	GetRegistryFormat() common.RegistryFormat
+
+	// ListTypes lists types in this registry whose string values conform to the
+	// supplied regular expression, or all types, if the regular expression is nil.
+	ListTypes(regex *regexp.Regexp) ([]Type, error)
+	// GetDownloadURLs returns the URLs required to download the type contents.
+	GetDownloadURLs(t Type) ([]*url.URL, error)
+}
+
+// GithubRegistry abstracts a registry that resides in a Github repository.
+type GithubRegistry interface {
+	Registry // A GithubRegistry is a Registry.
+	// GetRegistryOwner returns the owner name for this registry
+	GetRegistryOwner() string
+	// GetRegistryRepository returns the repository name for this registry.
+	GetRegistryRepository() string
+	// GetRegistryPath returns the path to the registry in the repository.
+	GetRegistryPath() string
 }
 
-// Registry abstracts a registry that holds templates, which can be
-// used in a Deployment Manager configurations. There can be multiple
-// implementations of a registry. Currently we support Deployment Manager
-// github.com/kubernetes/application-dm-templates
-// and helm packages
-// github.com/helm/charts
-//
 type Type struct {
 	Collection string
 	Name       string
-	Version    string
+	version    SemVer
+}
+
+// NewType initializes a type
+func NewType(collection, name, version string) (Type, error) {
+	result := Type{Collection: collection, Name: name}
+	err := result.SetVersion(version)
+	return result, err
+}
+
+// NewTypeOrDie initializes a type and panics if initialization fails
```

Comment:
It's either client or server. Currently runs in dm CLI. Up to the caller to use the right method (e.g., go has panic, and it's up to the caller to decide when to use it or not). Useful in this form for initializing test data structs.